mclagsyncd/mclaglink.cpp

```diff
@@ -0,0 +1,814 @@
+{
+    struct sockaddr_in client_addr;
+    socklen_t client_len;
+
+    m_connection_socket = ::accept(m_server_socket, (struct sockaddr *)&client_addr,
+                                   &client_len);
+    if (m_connection_socket < 0)
+        throw system_error(errno, system_category());
+
+    SWSS_LOG_NOTICE("New connection accepted from: %s", inet_ntoa(client_addr.sin_addr));
+}
+
+int MclagLink::getFd()
+{
+    return m_connection_socket;
+}
+
+void MclagLink::readData()
+{
+    mclag_msg_hdr_t *hdr = NULL;
+    size_t msg_len = 0;
+    size_t start = 0, left = 0;
+    ssize_t read = 0;
+    ssize_t write = 0;
+    char * msg = NULL;
+
+    read = ::read(m_connection_socket, m_messageBuffer + m_pos, m_bufSize - m_pos);
+    if (read == 0)
+        throw MclagConnectionClosedException();
+    if (read < 0)
+        throw system_error(errno, system_category());
+    m_pos += (uint32_t)read;
+
+    while (true)
+    {
+        hdr = (mclag_msg_hdr_t *)(m_messageBuffer + start);
+        left = m_pos - start;
+        if (left < MCLAG_MSG_HDR_LEN)
+            break;
+
+        msg_len = mclag_msg_len(hdr);
+        if (left < msg_len)
+            break;
+
+        if (!mclag_msg_ok(hdr, left))
+            throw system_error(make_error_code(errc::bad_message), "Malformed MCLAG message received");
+
+        msg = ((char*)hdr) + MCLAG_MSG_HDR_LEN;
+
+        switch (hdr->msg_type)
+        {
+            case MCLAG_MSG_TYPE_PORT_ISOLATE:
+                mclagsyncd_set_port_isolate(msg);
+                break;
+
+            case MCLAG_MSG_TYPE_PORT_MAC_LEARN_MODE:
+                mclagsyncd_set_port_mac_learn_mode(msg);
+                break;
+
+            case MCLAG_MSG_TYPE_FLUSH_FDB:
+                mclagsyncd_set_fdb_flush();
+                break;
+
+            case MCLAG_MSG_TYPE_FLUSH_FDB_BY_PORT:
+                mclagsyncd_set_fdb_flush_by_port(msg);
+                break;
+
+            case MCLAG_MSG_TYPE_SET_INTF_MAC:
+                mclagsyncd_set_intf_mac(msg);
+                break;
+
+            case MCLAG_MSG_TYPE_SET_FDB:
+                mclagsyncd_set_fdb_entry(msg, (int)(hdr->msg_len - sizeof(mclag_msg_hdr_t)));
+                break;
+
+            case MCLAG_MSG_TYPE_GET_FDB_CHANGES:
+                write = mclagsyncd_get_fdb_changes(m_messageBuffer_send);
+                if (write == 0)
+                    throw MclagConnectionClosedException();
+                if (write < 0)
+                    throw system_error(errno, system_category());
+                break;
+
+            default:
+                break;
+        }
+
+        start += msg_len;
+    }
+
+    memmove(m_messageBuffer, m_messageBuffer + start, m_pos - start);
+    m_pos = m_pos - (uint32_t)start;
+    return;
+}
+
+void MclagLink::mclag_connection_lost_handle_port_isolate()
+{
+    string acl_name = "mclag";
+
+    p_acl_tbl->del(acl_name);
+
+    return;
+}
+
+void MclagLink::mclag_connection_lost_handle_port_learn_mode()
```

Comment:
What is the intended usage of this API and the MclagLink::mclag_connection_lost_handle_port_isolate()? Do you plan to revert all changes to hardware when mclagsyncd's connection to ICCPd fails? If yes, please elaborate more on the overall scheme and traffic impact implication.

Reply:
@lisahnguyen, We will delete the unused codes.